### ISSUE TITLE
How to get hold of the Text Highlight Color in HTML?

### ISSUE THREAD

**Issue by hamzamansoor15** [mwilliamson/mammoth.js#395]
I want to get the text that is Highlight by the Word's Text Highlight Color Tool.

I ran `await mammoth.convertToHtml({ path: inputDoc });` on the Docx file which converted the content in html. That worked fine.
Now while reading the text of the node in HTML, Highlighted part is also being shown as regular text, there is no identification that which part of the color was highlighted?

Should there be some something like `<mark>highlighted text</mark>`?

Does anyone know the solution to my issue?

<img width="1792" alt="Screenshot 2024-03-29 at 2 52 22 AM" src="https://github.com/mwilliamson/mammoth.js/assets/30665929/28d05d44-c8f8-4d12-869a-b4cb26ba414b">

**Comment by jan-schweiger** [comment]
Any updates on this? @mwilliamson

That would be really great.

**Comment by mwilliamson** [comment]
Could you provide a minimal example document, the expected HTML, and the actual HTML you're getting?

**Comment by jan-schweiger** [comment]
[Document.docx](https://github.com/user-attachments/files/15775359/Document.docx)

Thank you for your quick reply @mwilliamson .

Sure, if you convert a docx file with a highlighted text, it would be great to somehow extract the background color or just to get the information that the text is highlighted.

The html I currently get is:
```
<p>This is a highlighted text.</p>
```

Therefore this information is lost. It would be great, if the information is somehow preserved. Here are three possible examples that would be sufficient for our use case:
### Option 1:
```
<p>This is a <span style="background-color: ##FFFF00;">highlighted text</span>.</p>
```

### Option 2:
```
<p>This is a <span background-color="##FFFF00">highlighted text</span>.</p>
```

### Option 3 (No color):
```
<p>This is a <hi>highlighted text</hi>.</p>
```

I know Mammoth works with styles, however in most use cases the Word files don't use custom styles for highlights. In our use case the Word files come from the customer, so we have even less control.

It would be great to somehow get this information. Thank you!

**Comment by jan-schweiger** [comment]
Thanks a lot for helping us @mwilliamson!

I have tried to implement your solution. Unfortunately, I cannot find the highlighting in the runs.

Here is my code that outputs the runs:
```
const transformRun = element => {
  console.log(element);
  return element;
}

const transformElement = element => {
  if (element.children) {
    var children = element.children.map(transformElement);
    element = {...element, children: children};
  }

  if (element.type === "run")
    element = transformRun(element);
  return element;
}
```

I get the highlighted text part as separate run. However, there is no attribute that indicates that this particular run is highlighted or not.
```
{
  type: 'run',
  children: [ { type: 'text', value: 'highlighted text' } ],
  styleId: null,
  styleName: null,
  isBold: false,
  isUnderline: false,
  isItalic: false,
  isStrikethrough: false,
  isAllCaps: false,
  isSmallCaps: false,
  verticalAlignment: 'baseline',
  font: null,
  fontSize: null
}
```

Could you maybe tell me what I missed? Is there some code, that may inspire us to find a solution?

Thanks again,
Jan

**Comment by mwilliamson** [comment]
If you want to use a document transform, then support for parsing highlighted runs will need to be added to the parser.

**Comment by mwilliamson** [comment]
To clarify, my earlier message was intended to try to work out what your use case is and what solutions would work best, rather than saying that you could do so with the code as-is.

**Comment by jan-schweiger** [comment]
Thank you for your reply @mwilliamson.

From my understanding it seems as extending the parser requires forking your repository, right? If that's the case that is quite unfortunate.

We will try to implement all other features first and then come back to this issue. Thank you for your help.

**Comment by mwilliamson** [comment]
I think adding support for reading highlights from the document is fine to go into Mammoth proper, but I'm just trying to understand your use case better to work out which of the options would work for you.

**Comment by jan-schweiger** [comment]
Sounds great @mwilliamson,

**Our use case:**
We provide a Tiptap-powered web-editor to our end users. We are currently building a Word import with your plugin.

Therefore, It would be great to have a custom html-tag that surrounds highlighted text in Word.

E.g.
```
<p>This is a <hi>highlighted text</hi>.</p>
```

If we would be able to get the background-color it would be even better.

```
<p>This is a <hi color="#ff0000">highlighted text</hi>.</p>
```

**Comment by mwilliamson** [comment]
`master` should now [support style mappings for highlights](https://github.com/mwilliamson/mammoth.js?tab=readme-ov-file#highlight), if you'd like to give it a try.

If you're using Mammoth in a commercial product, then feel free to make a [one-off](https://ko-fi.com/mwilliamson) or [recurring donation](https://liberapay.com/mwilliamson) to support development.

**Comment by jan-schweiger** [comment]
@mwilliamson Thanks a lot for your help.

I became a recurring patron. If we launch our product with your plugin, our company might sponsor you.

When will you include this feature in the next npm release?

Thanks, Jan

**Comment by mwilliamson** [comment]
> I became a recurring patron.

Thanks!

> When will you include this feature in the next npm release?

It should be good to go now, but it'd be useful to get any feedback on whether the implementation actually works for anyone's use case.

**Comment by jan-schweiger** [comment]
Hi @mwilliamson, 

I tried the latest version from Github and it looks great.

```
  const options = {
    styleMap: [
      "highlight[color='yellow'] => mark:fresh"
       // I also successfully tried and used "highlight => mark:fresh" to map all highlights
    ]
  };
```

Thank you for your great work!!

**Comment by mwilliamson** [comment]
Thanks, this has now been released as part of 1.8.0.
